router/internal/eventorder/eventorder.go

```diff
@@ -0,0 +1,248 @@
+package eventorder
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/rudderlabs/rudder-server/jobsdb"
+)
+
+var ErrUnsupportedState = errors.New("unsupported state")
+
+// NewBarrier creates a new properly initialized Barrier
+func NewBarrier(abortConcurrencyLimit int) *Barrier {
+	return &Barrier{
+		concurrencyLimit: abortConcurrencyLimit,
+		barriers:         make(map[string]*barrierInfo),
+	}
+}
+
+type Barrier struct {
```

Comment:
We could have a comment here explaining what a `Barrier` is for.

Reply:
yes, how did I miss it :)